keps/sig-node/20191002-seccomp-custom-profiles.md

```diff
@@ -0,0 +1,351 @@
+if !ok {
+  ...
+}
+if !isValidJson(profileData){
+  ...
+}
+
+return []dockerOpt{{"seccomp", profileData, ""}}, nil
+```
+
+### Feature Gate
+
+A new feature gate will be created to enable the use of Seccomp ConfigMaps, as per below:
+
+`--feature-gate=SeccompConfigMaps=true`
+
+The proposed default value for this feature gate will change through the [Graduation Process](#graduation-criteria).
+
+### Test Plan
+
+Seccomp already has E2E tests, but the tests are guarded by the [Feature:Seccomp] tag and not run in the standard test suites.
+
+Prior to [being marked GA](20190717-seccomp-ga.md), the feature tag will be removed from the seccomp tests.
+
+Unit tests will be created for both kubelet and API server to account for configmap profiles, taking into account scenarios mentioned at the [usage scenarios](#usage-scenarios) section.
+
+E2E tests will be required to ensure profile changes won't impact running containers.
+
+### Upgrade / Downgrade Strategy
+
+No upgrade changes required - both localhost and configmap profiles will be supported in parallel.
+
+### Version Skew Strategy
+
+Error handling for users trying to use the new `ConfigMap` based profiles on older versions of either Kubelet or API server is already implemented. Messages will differ depending on what component does not support the new feature:
+
+- **Not Supported at API Server** (i.e. before implementation of this proposal):
+Pod creation will fail with the error message `must be a valid seccomp profile`. 
+
+- **Supported at API Server but not at Kubelet** (i.e. Kubelet running on older version):
+Pod creation will fail with the error message `unknown seccomp profile option: configmap/config-name`. 
+
+The error message above may change once seccomp [becomes GA](20190717-seccomp-ga.md).
+
+### Graduation Criteria
+
+
+##### Alpha
+
+- API changes to map Built-in profiles to an additional `Kubernetes` type.
+- Feature gate is disabled by default (i.e. `--feature-gate=SeccompConfigMaps=false`).
+- Provide alpha level documentation.
+- Unit tests.
+
+##### Alpha -> Beta Graduation
+
+- Feature gate is enabled by default (i.e. `--feature-gate=SeccompConfigMaps=true`).
+- Approach community for feedback.
+- Provide beta level documentation.
+- E2E tests.
+
+##### Beta -> GA Graduation
+
+- Provide comprehensive documentation.
+- Feature gate is removed.
+- Potential changes based on community feedback.
+
+
+## Implementation History
+- 2019-10-02: Initial KEP
+- 2019-10-15: Minor changes
+- 2019-10-17: Add alternative to use CRD instead of ConfigMap
+- 2019-10-27: Add improvements to Design, Version Skew Strategy and Graduation
+- 2019-10-29: Add improvements to alternatives and multiple files support
+
+
+## Alternatives
+
+### Use CRD instead of ConfigMap
+
+The decision to use `ConfigMap` was to avoid unnecessary complexity. However, this can be later reviewed through the Graduation process. The reasons considered whilst assessing the two different approaches were:
```

Comment:
Honestly spoken I'm totally in favor of a CRD. I don't think that users write JSON seccomp profiles by hand. It's not possible to validate them upon creation and we cannot guarantee that a typo in a profile opens-up security issues.

Reply:
I agree with @saschagrunert .

Reply:
@saschagrunert @evrardjp I would also prefer the CRD approach, but ended up going for the ConfigMap to minimise changes needed so we could push this over the line sooner. 

Once the other details are ironed out, we could potentially consider moving to a CRD for Beta/GA. What do you think?

Reply:
Sounds like a plan. I think we can iterate on this and keep the CRD out of scope for now (see my other comments)